File: 32. Longest Valid Parentheses Hard/solution.ts
function longestValidParentheses(s: string): number {
    let stack: number[] = [-1];
    let result = 0;

    for (let i = 0; i < s.length; i++) {
        if (s.charAt(i) == "(") {
            stack.push(i);
        } else {
            stack.pop();
            if (stack.length == 0) {
                stack.push(i);
            } else {
                result = Math.max(result, i - stack[stack.length-1]);
            }
        }
    }

    return result;
}

console.log(longestValidParentheses("(()"));
console.log(longestValidParentheses(")()())"));
console.log(longestValidParentheses(""));
console.log(longestValidParentheses("((()))"));
console.log(longestValidParentheses("((()))()"));
console.log(longestValidParentheses("(()()())()((()))"));

File: 32. Longest Valid Parentheses Hard/solution.py
def longest_valid_parentheses(s: str) -> int:
    stack = [-1]
    result = 0
    for i, c in enumerate(s):
        if c == "(":
            stack.append(i)
        else:
            stack.pop()
            if not stack:
                stack.append(i)
            else:
                result = max(result, i - stack[-1])
    return result

print(longest_valid_parentheses("(()"))
print(longest_valid_parentheses(")()())"))
print(longest_valid_parentheses(""))
print(longest_valid_parentheses("((()))"))
print(longest_valid_parentheses("((()))()"))
print(longest_valid_parentheses("(()()())()((()))"))